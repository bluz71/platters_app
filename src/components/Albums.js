import React, { Component } from 'react';
import axios from 'axios';
import { Col, PageHeader } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import FontAwesome from 'react-fontawesome';
import pluralize from 'pluralize';
import numeral from 'numeral';
import queryString from 'query-string';
import _ from 'lodash';
import { VelocityTransitionGroup } from 'velocity-react';
import NProgress from 'nprogress';
import '../styles/Albums.css';
import { API_HOST } from '../config';
import Search from './Search';
import Paginator from './Paginator';

const ALBUMS_ENDPOINT = `${API_HOST}/albums.json`;

class Albums extends Component {
  constructor(props) {
    super(props);

    // params is not React state since we do not want to re-render when it
    // changes; just use an instance variable instead.
    this.params = {};
    this.loaded = false;
    this.searching = false;

    this.state = {
      albums: [],
      pagination: {},
      error: null
    };
  }

  componentDidMount() {
    window.onpopstate = this.handlePopState;
    this.applyState();
  }

  componentWillUnmount() {
    window.onpopstate = null;
  }

  shouldComponentUpdate(nextProps) {
    // Location change indicates either a direct 'this.props.history.push' or
    // back/forward in a browser. In both circumstances we do NOT want to
    // re-render since a GET for a new set of resources will immediately occur
    // after which will result in a re-render. Basically we only want to render
    // once upon a state change, not twice, hence we can safely ignore location
    // changes.
    return _.isEqual(this.props.location, nextProps.location);
  }

  handlePopState = (event) => {
    event.preventDefault();
    this.applyState();
  }

  handlePageChange = (page) => {
    const newParams = { ...this.params, page };
    this.applyParams(newParams);
  }

  handleAll = () => {
    const searching = this.searching;
    this.searching = false;
    const newParams = {};
    if (!this.applyParams(newParams) && searching) {
      // Hide the search form if it was previously visible.
      this.forceUpdate();
    }
  }

  handleYear = (year) => {
    const newParams = { year };
    this.applyParams(newParams);
  }

  handleGenre = (genre) => {
    const newParams = { genre };
    this.applyParams(newParams);
  }

  handleLetter = (letter) => {
    const newParams = _.omit(this.params, ['page', 'search']);
    newParams.letter = letter;
    this.applyParams(newParams);
  }

  handleRandom = () => {
    const newParams = { random: true };
    this.applyParams(newParams);
  }

  handleFilterVisibility = () => {
    console.log('In Filter Visibility');
  }

  handleSearchVisibility = () => {
    this.searching = !this.searching;
    this.forceUpdate();
  }

  handleSearchChange = (term) => {
    const newParams = term ? { search: term} : {};
    this.applyParams(newParams);
  }

  handleSearchSubmit = (event) => {
    event.preventDefault();
    const term = event.target.elements.search.value;
    const newParams = term ? { search: term} : {};
    this.applyParams(newParams);
  }

  // Apply state for back and forward transistion into/outof/within this
  // component.
  applyState() {
    if (this.props.location) {
      this.params = this.props.location.state || {};
    }
    else {
      this.params = {};
    }
    this.getAlbums(false);
  }

  // Apply paramaters and update resources only if they are changed.
  applyParams(newParams) {
    if (_.isEqual(this.params, newParams) && !newParams.hasOwnProperty('random')) {
      // Nothing to do, new params have not changed, just return.
      // Note, if random is set then ignore this opt-out since it will always
      // generate a new set of albums.
      return false;
    }

    this.params = newParams;
    this.props.history.push('/albums', this.params);
    this.getAlbums();
    return true;
  }

  albumsURL() {
    const params = queryString.stringify(this.params);

    return params.length > 0 ? `${ALBUMS_ENDPOINT}?${params}` : ALBUMS_ENDPOINT;
  }

  getAlbums(scrollToTop = true) {
    axios.get(this.albumsURL())
      .then(response => {
        if (!this.loaded) {
          this.loaded = true;
          NProgress.done();
        }
        this.setState({
          albums: response.data.albums,
          pagination: response.data.pagination,
          error: null
        });
        if (scrollToTop) {
          window.scrollTo(0, 0);
        }
      })
      .catch(error => {
        this.setState({
          error: error
        });
      });
  }

  letterActivity(letter) {
    if (this.params.letter === letter) {
      return 'active';
    }
  }

  renderHeader() {
    const count = this.state.pagination.total_count;
    const albumsCount = numeral(count).format('0,0');
    const genre = this.params.hasOwnProperty('genre') ? this.params.genre : '';
    const year = this.params.hasOwnProperty('year') ? ` from ${this.params.year}` : '';

    return (
      <PageHeader>
        Albums {this.loaded && <small>({albumsCount} {genre} {pluralize('Album', count)}{year})</small>}
      </PageHeader>
    );
  }

  showLettersFilter() {
    if (this.searching || this.params.hasOwnProperty('random')) { 
       return false;
    }
    else {
      return true;
    }
  }

  renderSearch() {
    return (
      <VelocityTransitionGroup
        enter={{
          animation: 'slideDown',
          duration: 250,
          complete: () => {
            this.search.focusSearchInput();
          }
        }}
        leave={{animation: 'slideUp', duration: 250}}>
        {this.searching &&
          <Search
            placeholder="Search Artists..."
            onSearchChange={this.handleSearchChange}
            onSearchSubmit={this.handleSearchSubmit}
            ref={(search) => this.search = search}
          />}
      </VelocityTransitionGroup>
    );
  }

  renderFilters() {
    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');

    return (
      <div>
        <div className="filters">
          <ul className="pagination pagination-sm">
            <li onClick={this.handleAll}><a>All</a></li>
            {this.showLettersFilter() && letters.map(letter => <li onClick={() => this.handleLetter(letter)} key={letter} className={this.letterActivity(letter)}><a>{letter}</a></li>)}
            <li onClick={this.handleRandom}><FontAwesome name="random" /></li>
            <li onClick={this.handleFilterVisibility}><FontAwesome name="filter" /></li>
            <li onClick={this.handleSearchVisibility}><FontAwesome name="search" /></li>
          </ul>
        </div>
        {this.renderSearch()}
      </div>
    );
  }

  renderAlbums() {
    if (!this.loaded) {
      NProgress.start();
    }

    if (this.state.error) {
      return <h4>Error retrieving albums</h4>;
    }

    if (this.loaded && this.state.albums.length === 0) {
      return <h4>No matching albums</h4>;
    }

    return (
      this.state.albums.map(album =>
        <Col key={album.id} md={6}>
          <div className="Album">
            <h2>
              <Link to={`/artist/${album.artist_slug}/album/${album.album_slug}`}>{album.title}</Link>
            </h2>
            <h3>
              by <Link to={`/artist/${album.artist_slug}`}>{album.artist}</Link> <small>({album.tracks_count} {pluralize('Track', album.tracks_count)})</small>
            </h3>
            <div className="icon">
              <a onClick={() => this.handleYear(album.year)}><FontAwesome name="calendar" /> {album.year}</a>
              <a onClick={() => this.handleGenre(album.genre)}><FontAwesome name="tag" className="spacer-left-xs" /> {album.genre}</a>
              <FontAwesome name="comment-o" className="spacer-left-xs" /> {album.comments_count}
            </div>
            <Link to={`/artist/${album.artist_slug}/album/${album.album_slug}`}>
              <img className="img-responsive" src={album.cover_url} alt={album.title} />
            </Link>
            <ul>
              {album.tracks.map((track, index) => <li key={index}>{track}</li>)}
            </ul>
          </div>
        </Col>
      )
    );
  }

  renderPaginator() {
    if (this.state.pagination.total_pages > 1) {
      return (
        <Col md={8}>
          <Paginator pagination={this.state.pagination} onPageChange={this.handlePageChange} />
        </Col>
      );
    }
  }

  render() {
    return (
      <div className="Albums">
        {this.renderHeader()}
        {this.renderFilters()}
        {this.renderAlbums()}
        {this.renderPaginator()}
      </div>
    );
  }
}

export default Albums;
