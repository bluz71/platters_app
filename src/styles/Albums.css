.Albums h4 {
  text-align: center;
}

.Albums .col-md-6 {
  padding: 0 10px;
}

.Albums .filters {
  text-align: center;
}

.Albums .filters .pagination {
  margin: 0 0 0.5em;
}

.Albums .filters .pagination li {
  cursor: pointer;
}

.Album {
  background-color: #f8f8f8;
  border: 1px solid #eee;
  border-radius: 4px;
  height: 18.5em;
  margin-top: 10px;
  margin-bottom: 8px;
}

.Album h2 {
  padding: 0 0.25em 0.1em 0.25em;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 100%;
  text-align: center;
}

.Album h3 {
  margin-top: 0.25em;
  text-align: center;
  color: #111;
  font-size: 1em;
}

.Album h3 small {
  font-size: 0.8em;
}

.Album a {
  color: #515151;
  text-decoration: none;
}

.Album a:hover {
  color: #3c9a5f;
}

.Album .icon {
  text-align: center;
  font-size: 0.75em;
  color: #515151;
  margin-bottom: 0.5em;
}

.Album .icon a {
  cursor: pointer;
}

.Album img {
  height: 125px;
  width: 125px;
  float: left;
  margin: 15px 15px 15px 20px;
  border: 1px solid #777;
  border-radius: 4px;
}

.Album img:hover {
  border: 1px solid #3c9a5f;
}

.Album ul {
  display: inline;
  font-size: 0.75em;
  color: #515151;
}

.Album ul li {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  list-style-type: none;
}

.Albums .loading-spinner {
  text-align: center;
  font-size: 2em;
  color: #777;
}
