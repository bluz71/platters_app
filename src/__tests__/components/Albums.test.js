import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Albums from '../../components/Albums';

const flushPromises = () => (
  new Promise(resolve => setImmediate(resolve))
);

describe('<Albums />', () => {
  it('renders without crashing', () => {
    shallow(<Albums />);
  });

  it('renders div.Albums', () => {
    const albums = shallow(<Albums />);
    expect(albums.find('div.Albums').length).toEqual(1);
  });

  it('renders default snapshot', async () => {
    const wrapper = shallow(<Albums />);
    await flushPromises();
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
