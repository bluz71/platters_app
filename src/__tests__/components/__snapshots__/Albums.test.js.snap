// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Albums /> renders default snapshot 1`] = `
<div
  className="Albums"
>
  <PageHeader
    bsClass="page-header"
  >
    Albums 
    <small>
      (
      6
       
       
      Albums
      )
    </small>
  </PageHeader>
  <div
    className="filters"
  >
    <ul
      className="pagination pagination-sm"
    >
      <li
        onClick={[Function]}
      >
        <a>
          All
        </a>
      </li>
      <li
        key="A"
        onClick={[Function]}
      >
        <a>
          A
        </a>
      </li>
      <li
        key="B"
        onClick={[Function]}
      >
        <a>
          B
        </a>
      </li>
      <li
        key="C"
        onClick={[Function]}
      >
        <a>
          C
        </a>
      </li>
      <li
        key="D"
        onClick={[Function]}
      >
        <a>
          D
        </a>
      </li>
      <li
        key="E"
        onClick={[Function]}
      >
        <a>
          E
        </a>
      </li>
      <li
        key="F"
        onClick={[Function]}
      >
        <a>
          F
        </a>
      </li>
      <li
        key="G"
        onClick={[Function]}
      >
        <a>
          G
        </a>
      </li>
      <li
        key="H"
        onClick={[Function]}
      >
        <a>
          H
        </a>
      </li>
      <li
        key="I"
        onClick={[Function]}
      >
        <a>
          I
        </a>
      </li>
      <li
        key="J"
        onClick={[Function]}
      >
        <a>
          J
        </a>
      </li>
      <li
        key="K"
        onClick={[Function]}
      >
        <a>
          K
        </a>
      </li>
      <li
        key="L"
        onClick={[Function]}
      >
        <a>
          L
        </a>
      </li>
      <li
        key="M"
        onClick={[Function]}
      >
        <a>
          M
        </a>
      </li>
      <li
        key="N"
        onClick={[Function]}
      >
        <a>
          N
        </a>
      </li>
      <li
        key="O"
        onClick={[Function]}
      >
        <a>
          O
        </a>
      </li>
      <li
        key="P"
        onClick={[Function]}
      >
        <a>
          P
        </a>
      </li>
      <li
        key="Q"
        onClick={[Function]}
      >
        <a>
          Q
        </a>
      </li>
      <li
        key="R"
        onClick={[Function]}
      >
        <a>
          R
        </a>
      </li>
      <li
        key="S"
        onClick={[Function]}
      >
        <a>
          S
        </a>
      </li>
      <li
        key="T"
        onClick={[Function]}
      >
        <a>
          T
        </a>
      </li>
      <li
        key="U"
        onClick={[Function]}
      >
        <a>
          U
        </a>
      </li>
      <li
        key="V"
        onClick={[Function]}
      >
        <a>
          V
        </a>
      </li>
      <li
        key="W"
        onClick={[Function]}
      >
        <a>
          W
        </a>
      </li>
      <li
        key="X"
        onClick={[Function]}
      >
        <a>
          X
        </a>
      </li>
      <li
        key="Y"
        onClick={[Function]}
      >
        <a>
          Y
        </a>
      </li>
      <li
        key="Z"
        onClick={[Function]}
      >
        <a>
          Z
        </a>
      </li>
    </ul>
  </div>
  <Col
    bsClass="col"
    componentClass="div"
    key="1"
    md={6}
  >
    <div
      className="Album"
    >
      <h2>
        <Link
          replace={false}
          to="/artist/abc_artist/album/abc"
        >
          ABC
        </Link>
      </h2>
      <h3>
        by 
        <Link
          replace={false}
          to="/artist/abc_artist"
        >
          ABC-Artist
        </Link>
         
        <small>
          (
          15
           
          Tracks
          )
        </small>
      </h3>
      <div
        className="icon"
      >
        <a
          onClick={[Function]}
        >
          <FontAwesome
            name="calendar"
          />
           
          2008
        </a>
        <a
          onClick={[Function]}
        >
          <FontAwesome
            className="spacer-left-xs"
            name="tag"
          />
           
          Rock
        </a>
        <FontAwesome
          className="spacer-left-xs"
          name="comment-o"
        />
         
        42
      </div>
      <Link
        replace={false}
        to="/artist/abc_artist/album/abc"
      >
        <img
          alt="ABC"
          className="img-responsive"
          src="http://localhost:3000/uploads/album/cover/1/small_abc--abc.jpg"
        />
      </Link>
      <ul>
        <li
          key="0"
        >
          1. First
        </li>
        <li
          key="1"
        >
          2. Second
        </li>
        <li
          key="2"
        >
          3. Third
        </li>
        <li
          key="3"
        >
          4. Fourth
        </li>
        <li
          key="4"
        >
          5. Fifth
        </li>
        <li
          key="5"
        >
          6. Sixth
        </li>
        <li
          key="6"
        >
          ...
        </li>
      </ul>
    </div>
  </Col>
  <Col
    bsClass="col"
    componentClass="div"
    key="2"
    md={6}
  >
    <div
      className="Album"
    >
      <h2>
        <Link
          replace={false}
          to="/artist/def_artist/album/def"
        >
          DEF
        </Link>
      </h2>
      <h3>
        by 
        <Link
          replace={false}
          to="/artist/def_artist"
        >
          DEF-Artist
        </Link>
         
        <small>
          (
          27
           
          Tracks
          )
        </small>
      </h3>
      <div
        className="icon"
      >
        <a
          onClick={[Function]}
        >
          <FontAwesome
            name="calendar"
          />
           
          2015
        </a>
        <a
          onClick={[Function]}
        >
          <FontAwesome
            className="spacer-left-xs"
            name="tag"
          />
           
          Classical
        </a>
        <FontAwesome
          className="spacer-left-xs"
          name="comment-o"
        />
         
        44
      </div>
      <Link
        replace={false}
        to="/artist/def_artist/album/def"
      >
        <img
          alt="DEF"
          className="img-responsive"
          src="http://localhost:3000/uploads/album/cover/978/small_def--def.jpg"
        />
      </Link>
      <ul>
        <li
          key="0"
        >
          1. One
        </li>
        <li
          key="1"
        >
          2. Two
        </li>
      </ul>
    </div>
  </Col>
  <Col
    bsClass="col"
    componentClass="div"
    md={8}
  >
    <Paginator
      onPageChange={[Function]}
      pagination={
        Object {
          "current_page": 1,
          "next_page": 2,
          "prev_page": null,
          "total_count": 6,
          "total_pages": 2,
        }
      }
    />
  </Col>
</div>
`;
